### PR TITLE
fix: deposit flow

### PR DESCRIPTION
**Pull request by Schlagonia**
## Description

- Combine internal deposit and mint functionality into 1 function.
- Remove `issue_shares_for_amount` in order to use consistent logic for both routes.
- Still allow deposits AND mint to work if totalSupply == 0 but totalAssets > 0. Just make PPS == 1.

This can happen due to the profit unlocking if all shares are redeemed while some profit is still unlocking, at the end of the unlock time totalAssets >0 but totalSupply == 0.
- Still return 0 in convertToShares if totalSupply > 0 but totalAssets == 0. Meaning a full loss.
- Optimize to not cache totalAssets_ until needed
Fixes # (issue)

## Checklist

- [x] I have run vyper and solidity linting
- [x] I have run the tests on my machine
- [x] I have followed commitlint guidelines
- [x] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

**Review by fp-crypto — COMMENTED** [review]
More eyes are better

File: contracts/VaultV3.vy [hunk 1]
```diff
@@ -662,52 +638,16 @@ def _max_withdraw(
...
-def _deposit(sender: address, recipient: address, assets: uint256) -> uint256:
+def _deposit(sender: address, recipient: address, assets: uint256, shares: uint256):
     """
-    Used for `deposit` calls to transfer the amount of `asset` to the vault, 
-    issue the corresponding shares to the `recipient` and update all needed 
+    Used for `deposit` and `mint` calls to transfer the amount of `asset` to the vault, 
+    issue the corresponding `shares` to the `recipient` and update all needed 
     vault accounting.
     """
-    assert self.shutdown == False # dev: shutdown
```
Comment (wavey0x) [inline contracts/VaultV3.vy:671]: why remove this?

**Review by wavey0x — APPROVED** [review]

File: contracts/VaultV3.vy (hunk 1 above)
Comment (Schlagonia) [reply contracts/VaultV3.vy:671]: It is unnecesary.

When the vault is shutdown the deposit limit module is always set to address(0) and deposit limit to 0. https://github.com/yearn/yearn-vaults-v3/blob/411cae31a2e7cdae49364bdbfa3f601667ef141e/contracts/VaultV3.vy#L1772

And cannot be changed anymore. 

So _max_deposit() will always return 0, which will revert on the assertions

File: contracts/VaultV3.vy [hunk 2]
```diff
@@ -662,52 +638,16 @@ def _max_withdraw(
...
-    assets: uint256 = self._convert_to_assets(shares, Rounding.ROUND_UP)
-
-    assert assets > 0, "cannot deposit zero"
     assert assets <= self._max_deposit(recipient), "exceed deposit limit"
-
+    assert assets > 0, "cannot deposit zero"
+    assert shares > 0, "cannot mint zero"
+ 
     # Transfer the tokens to the vault first.
     self._erc20_safe_transfer_from(self.asset, msg.sender, self, assets)
```
Comment (spalen0) [inline contracts/VaultV3.vy:652]: Why not use the input parameter `sender` instead of `msg.sender`?
Below in the event Deposit, parameter `sender` is used.

I would either remove the sender parameter and go with `msg.sender` in both places or use the `sender` address for both, to limit confusion.

File: tests/unit/vault/test_shares.py [hunk 3]
```diff
@@ -498,3 +498,90 @@ def test__mint_shares_with_zero_total_supply_positive_assets(
...
+
+    vault.process_report(strategy, sender=gov)
+
+    assert vault.totalAssets() == 0
+    assert vault.totalSupply() != 0
+
+    with ape.reverts("cannot mint zero"):
+        vault.deposit(amount, fish, sender=fish)
+
+    # shares should not be
```
Comment (spalen0) [inline tests/unit/vault/test_shares.py:553]: Shares should not be minted?

File: tests/unit/vault/test_shares.py [hunk 4]
```diff
@@ -498,3 +498,90 @@ def test__mint_shares_with_zero_total_supply_positive_assets(
...
+    assert vault.balanceOf(fish) == amount
+    assert vault.pricePerShare() == 0
+    assert vault.convertToShares(amount) == 0
+    assert vault.convertToAssets(amount) == 0
+    # assert vault.maxDeposit(fish) == 0
```
Comment (spalen0) [inline tests/unit/vault/test_shares.py:558]: `maxDeposit()` won't return 0 but `uint256.max`? https://github.com/yearn/yearn-vaults-v3/blob/60524bd4d80419672429ab7218d95e7b6e353375/contracts/VaultV3.vy#L529

File: tests/unit/vault/test_shares.py [hunk 5]
```diff
@@ -498,3 +498,90 @@ def test__mint_shares_with_zero_total_supply_positive_assets(
...
+
+    with ape.reverts("cannot deposit zero"):
+        vault.mint(amount, fish, sender=fish)
+
+    # shares should not be
+    assert vault.balanceOf(fish) == amount
+    assert vault.pricePerShare() == 0
+    assert vault.convertToShares(amount) == 0
+    assert vault.convertToAssets(amount) == 0
+    # assert vault.maxMint(fish) == 0
```
Comment (spalen0) [inline tests/unit/vault/test_shares.py:587]: The same questions as above, for deposit flow.

File: contracts/VaultV3.vy (hunk 2 above)
Comment (Schlagonia) [reply contracts/VaultV3.vy:652]: good point

I believe it is actually cheaper to use msg.sender using a memory variable. And clearer so will remove the variable

File: tests/unit/vault/test_shares.py (hunk 4 above)
Comment (Schlagonia) [reply tests/unit/vault/test_shares.py:558]: yes, maxDeposit and maxMint wont be accurate currently in this scenario

Not sure any good way to fix it though without adding way to much complexity and extra gas.

Seems like such a minor edge case that it is not worth it.

File: tests/unit/vault/test_shares.py (hunk 5 above)
Comment (Schlagonia) [reply tests/unit/vault/test_shares.py]: maxMint will be accurate if there is a non-max uint deposit limit, since it will then do the full conversion

File: tests/unit/vault/test_shares.py (hunk 4 above)
Comment (Schlagonia) [reply tests/unit/vault/test_shares.py:558]: Realistically if this happens no-one should be trying to deposit and a manager should shutdown the vault anyway which will reduce the deposit limit.

File: tests/unit/vault/test_shares.py (hunk 3 above)
Comment (Schlagonia) [reply tests/unit/vault/test_shares.py]: https://github.com/yearn/yearn-vaults-v3/pull/209/commits/6a5ead1ffc7be7d4e4614574f6519b41996cd297

File: contracts/VaultV3.vy (hunk 2 above)
Comment (Schlagonia) [reply contracts/VaultV3.vy:652]: https://github.com/yearn/yearn-vaults-v3/pull/209/commits/5b7591662c845a3d141a715a78f3340d591cb38f

**Review by spalen0 — APPROVED** [review]
